address-space-controller/src/main/java/io/enmasse/controller/auth/OpenshiftCertProvider.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2018, EnMasse authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.enmasse.controller.auth;
+
+import io.enmasse.address.model.AddressSpace;
+import io.enmasse.address.model.KubeUtil;
+import io.enmasse.config.AnnotationKeys;
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.Service;
+import io.fabric8.openshift.client.OpenShiftClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OpenshiftCertProvider implements CertProvider {
+    private static final Logger log = LoggerFactory.getLogger(OpenshiftCertProvider.class);
+    private final OpenShiftClient client;
+    private final String namespace;
+
+    public OpenshiftCertProvider(OpenShiftClient client) {
+        this.client = client;
+        this.namespace = client.getNamespace();
+    }
+
+    @Override
+    public Secret provideCert(AddressSpace addressSpace, EndpointInfo endpointInfo) {
+        Secret secret = client.secrets().inNamespace(namespace).withName(endpointInfo.getCertSpec().getSecretName()).get();
+        if (secret == null) {
+            String serviceName = KubeUtil.getAddressSpaceServiceName(endpointInfo.getServiceName(), addressSpace);
+            Service service = client.services().withName(serviceName).get();
+            if (service != null && (service.getMetadata().getAnnotations() == null || service.getMetadata().getAnnotations().get(AnnotationKeys.OPENSHIFT_SERVING_CERT_SECRET_NAME) == null)) {
+                log.info("Adding service annotation to generate OpenShift cert");
+                client.services().withName(serviceName).edit()
+                        .editMetadata()
+                        .addToAnnotations(AnnotationKeys.OPENSHIFT_SERVING_CERT_SECRET_NAME, endpointInfo.getCertSpec().getSecretName())
+                        .endMetadata()
+                        .done();
```

Comment:
On the successful path you return null, which is different to the contracts implied by the other implementations.  Is the certificate generated by the time done() returns?  Should you be looking up the secret again and returning it?